ibm/service/power/resource_ibm_pi_virtual_serial_number.go

```diff
@@ -0,0 +1,386 @@
+	if _, ok := d.GetOk(Arg_PVMInstanceId); !ok && vsnArg == AutoAssign {
+		return diag.Errorf("cannot use '%s' unless %s is specified", AutoAssign, Arg_InstanceID)
+	}
+
+	serialString := ""
+	if vsnArg != AutoAssign {
+		vsn, err := client.Get(vsnArg)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		if v, ok := d.GetOk(Arg_Description); ok {
+			description := v.(string)
+			if description != *vsn.Description {
+				pvmInstanceId := *vsn.PvmInstanceID
+				if pvmInstanceId != "" {
+					updateBody := &models.UpdateServerVirtualSerialNumber{
+						Description: &description,
+					}
+					_, err := client.PVMInstanceUpdateVSN(pvmInstanceId, updateBody)
+					if err != nil {
+						return diag.FromErr(err)
+					}
+				} else {
+					updateBody := &models.UpdateVirtualSerialNumber{
+						Description: &description,
+					}
+					_, err := client.Update(vsnArg, updateBody)
+					if err != nil {
+						return diag.FromErr(err)
+					}
+				}
+			}
+		}
+		serialString = vsnArg
+	}
+
+	if pvmInstanceId, ok := d.GetOk(Arg_PVMInstanceId); ok {
+		pvmInstanceIdArg := pvmInstanceId.(string)
+		instanceClient := instance.NewIBMPIInstanceClient(ctx, sess, cloudInstanceID)
+		restartInstance, err := stopLparForVSNChange(ctx, instanceClient, pvmInstanceIdArg, d.Timeout(schema.TimeoutCreate))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		serialNumber := d.Get(Arg_Serial).(string)
+		addBody := &models.AddServerVirtualSerialNumber{
+			Serial: &serialNumber,
+		}
+		if v, ok := d.GetOk(Arg_Description); ok {
+			description := v.(string)
+			addBody.Description = description
+		}
+		err = client.PVMInstanceAttachVSN(pvmInstanceIdArg, addBody)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		_, err = isWaitForPIInstanceStopped(ctx, instanceClient, pvmInstanceIdArg, d.Timeout(schema.TimeoutCreate))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		if restartInstance {
+			err = startLparAfterVSNChange(ctx, instanceClient, pvmInstanceIdArg, d.Timeout(schema.TimeoutCreate))
+			if err != nil {
+				return diag.FromErr(err)
+			}
+		}
+
+		if vsnArg == AutoAssign {
+			vsns, err := client.GetAll(&pvmInstanceIdArg)
+			if err != nil {
+				return diag.FromErr(err)
+			}
+			serialString = *vsns[0].Serial
+
+		} else {
+			serialString = vsnArg
+		}
+	}
+
+	id := cloudInstanceID + "/" + serialString
+	d.SetId(id)
+
+	return resourceIBMPIVirtualSerialNumberRead(ctx, d, meta)
+}
+
+func resourceIBMPIVirtualSerialNumberRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(conns.ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	idArr, err := flex.IdParts(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudInstanceID := idArr[0]
+	client := instance.NewIBMPIVSNClient(ctx, sess, cloudInstanceID)
+
+	serialNumber := d.Get(Arg_Serial).(string)
+	vsn, err := client.Get(serialNumber)
+	if err != nil {
+		return diag.FromErr(err)
+	}
```

Comment:
```suggestion
	if err != nil {
		if strings.Contains(strings.ToLower(err.Error()), NotFound) {
			d.SetId("")
			return nil
		}
		return diag.FromErr(err)
	}
```
If it is deleted outside of TF